### PR TITLE
Make Datasets lazy-only

### PR DESCRIPTION
**Pull request by jianoaix**
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/pipeline_executor.py [hunk 1]
```diff
@@ -19,7 +19,9 @@ def pipeline_stage(fn: Callable[[], Dataset[T]]) -> Dataset[T]:
     # Force eager evaluation of all blocks in the pipeline stage. This
     # prevents resource deadlocks due to overlapping stage execution (e.g.,
     # task -> actor stage).
-    return fn().fully_executed()
+    ds = fn()
+    ds._plan.execute(force_read=True)
```
Comment (ericl) [inline python/ray/data/_internal/pipeline_executor.py:23]: Isn't this the same as fully_executed()? Do we not want to cache the output blocks here?

Comment (jianoaix) [reply python/ray/data/_internal/pipeline_executor.py:23]: The output blocks of transformations before the repeat/window are cached, but the outputs of pipeline are not. E.g. `ray.data.read_parquet().map_batches().window().map_batches().repeat(N)`, the results up to the first `map_batches()` are cached and reused, but the output of the second `map_batches()` are not.

Also I think the output blocks here by `pipeline_stage()` are not even the final outputs from the pipeline yet.

Comment (ericl) [reply python/ray/data/_internal/pipeline_executor.py:23]: Got it. Could we add a comment here to this effect?

Comment (jianoaix) [reply python/ray/data/_internal/pipeline_executor.py:23]: Yep, will do and polish once tests are fixed.

**Comment by jianoaix** [comment]
May revisit.

**Comment by jianoaix** [comment]
Since Dataset is already lazy and we already deprecated `ds.lazy()` API in Ray 2.0, closing this PR.

**Pull request closed without merging**